service-base/src/main/java/org/eclipse/hono/service/registration/BaseRegistrationService.java

```diff
@@ -304,6 +322,50 @@ public void assertRegistration(
         }).setHandler(resultHandler);
     }
 
+    private Future<RegistrationResult> updateDeviceLastViaIfNeeded(final String tenantId, final String deviceId,
+            final String gatewayId, final JsonObject deviceData) {
+        if (!isDeviceWithMultipleVias(deviceData)) {
+            return Future.succeededFuture();
+        }
+        final Future<RegistrationResult> lastViaUpdateFuture = Future.future();
+        updateLastVia(tenantId, deviceId, gatewayId, deviceData, lastViaUpdateFuture);
+        return lastViaUpdateFuture.map(res -> {
+            if (res != null && res.isError()) {
+                throw StatusCodeMapper.from(res);
+            }
+            return res;
+        });
+    }
+
+    private boolean isDeviceWithMultipleVias(final JsonObject deviceData) {
+        final Object viaObj = deviceData.getValue(PROPERTY_VIA);
+        return viaObj instanceof JsonArray && ((JsonArray) viaObj).size() > 1;
+    }
+
+    /**
+     * Updates device registration data and adds a 'last-via' property containing the given gateway identifier
+     * as well as the current date.
+     * <p>
+     * This method this called by this class' default implementation of <em>assertRegistration</em> for a device that
+     * has multiple gateway entries in its 'via' definition.
+     * <p>
+     * This default implementation simply invokes the given handler with a successful Future containing an empty result
+     * with status code 501 (Not Implemented).
+     * Subclasses need to override this method and provide a reasonable implementation in order for this
+     * class' default implementation of <em>assertRegistration</em> to work properly when using devices
+     * that have multiple gateway entries in their 'via' definition.
+     *
+     * @param tenantId The tenant id.
+     * @param deviceId The device id.
+     * @param gatewayId The gateway id.
+     * @param deviceData The current data associated with the device.
+     * @param resultHandler The handler to invoke with the result of the operation.
+     */
+    protected void updateLastVia(final String tenantId, final String deviceId, final String gatewayId,
+            final JsonObject deviceData, final Handler<AsyncResult<RegistrationResult>> resultHandler) {
```

Comment:
I do not quite understand why this method requires a `RegistrationResult` handler. FMPOV this method should simply return a `Future<Void` so that it can easily be composed with other futures ....

Reply:
My motivation to add the Handler as parameter (instead of returning a Future) was to keep the method signature consistent with the other methods of this class (addDevice/updateDevice/removeDevice). But I'm also ok with returning a Future instead.

Reply:
It's now changed to return `Future<Void>`.